File: CSharpRewrite/CSharpRewrite/Assignments/K08/Adventure.cs
using CSharpRewrite.Shared;

namespace CSharpRewrite.Assignments.K08;

public static class Adventure
{
    public static string GameLoop(int playerHp)
    {
        while (playerHp > 0)
        {
            Console.WriteLine($"あなたのHP：{playerHp}");
            Console.Write("奥に進みますか？（１：奥に進む　０．帰る）＞");

            switch (int.Parse(Console.ReadLine() ?? "0"))
            {
                case 0:
                    return "リレ〇ト！";
                default:

                    string? RunBattle()
                    {
                        var (hp, msg) = EngageBattle(playerHp);
                        if (hp.HasValue)
                            playerHp = hp.Value;
                        else if (msg != null)
                            return msg;
                        return null;
                    }

                    var result = RunBattle();
                    if (result != null)
                    {
                        Console.WriteLine(result);
                        return "ゲームオーバー！";
                    }
                    break;
            }
        }
        return "ゲームオーバー！";
    }

    private static int SelectAttack(int choice)
    {
        var rng = new Random();
        return choice switch
        {
            1 => 65 + rng.Next(0, 36),
            2 => 50 + rng.Next(0, 101),
            3 => 33 + rng.Next(0, 168),
            _ => SelectAttack(1)
        };
    }

    private static int DamagePlayer(int golemAttack, int playerHp)
    {
        Console.WriteLine("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
        Console.WriteLine($"ゴーレムがあなたを攻撃しました！攻撃値：{golemAttack}");
        return playerHp - golemAttack;
    }

    private static (int?, string?) BattleLoop(int golemLevel, Golem golem, int playerHp)
    {
        while (golem.Hp > 0)
        {
            Console.WriteLine($"ゴーレムLv.{golemLevel + 1} 残りHP：{golem.Hp}");
            Console.Write("武器を選択してください（１．攻撃　２．特技　３．魔法）＞");

            var baseDamage = SelectAttack(int.Parse(Console.ReadLine() ?? "0"));
            Console.WriteLine($"基礎ダメージは{baseDamage}です。");
            var actualDamage = baseDamage - golem.Defense < 0 ? 0 : baseDamage - golem.Defense;
            Console.WriteLine($"実際ダメージは{actualDamage}です。");

            if (actualDamage <= 0)
            {
                playerHp = DamagePlayer(golem.Attack, playerHp);
                if (playerHp <= 0)
                    return (null, "あなたはゴーレムに負けました！");
                Console.WriteLine($"あなたの残りHPは：{playerHp}");
            }
            else
            {
                golem.Hp = golem.Hp - actualDamage < 0 ? 0 : golem.Hp - actualDamage;
            }
        }
        
        Console.WriteLine($"ゴーレムLv.{golemLevel + 1}を倒した！");
        return (playerHp, null);
    }

    private static (int?, string?) EngageBattle(int playerHp)
    {
        var rng = new Random();
        var golemLevel = rng.Next(0, 10);
        var golem = new Golem
        {
            Hp = golemLevel * 50 + 100,
            Defense = golemLevel * 10 + 40,
            Attack = golemLevel * 10 + 30
        };
        
        Console.WriteLine($"ゴーレムLv.{golemLevel + 1}が現れた！");
        return BattleLoop(golemLevel, golem, playerHp);
    }
}

File: TypeScript/assignments/K08/Adventure.ts
import Golem from "../../shared/Golem.ts";
import { getRandomNumber } from "../../shared/Utility.ts";

interface ProgressResult {
    playerHp: number | null,
    message: string | null
}

function selectAttack(choice: number): number {
    switch (choice) {
        case 1:
            return 65 + getRandomNumber(36);
        case 2:
            return 50 + getRandomNumber(101);
        case 3:
            return 33 + getRandomNumber(168);
        default:
            return selectAttack(1);
    }
}

function damagePlayer(golemAttack: number, playerHp: number): number {
    console.log("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
    console.log(`ゴーレムがあなたを攻撃しました！攻撃値：${golemAttack}`);
    return playerHp - golemAttack;
}

function battleLoop(golemLevel: number, golem: Golem, playerHp: number): ProgressResult {
    while (golem.hp > 0) {
        console.log(`ゴーレムLv.${golemLevel + 1} 残りHP：${golem.hp}`)
        
        const baseDamage = selectAttack(Number.parseInt(prompt("武器を選択してください（１．攻撃　２．特技　３．魔法）＞") ?? "0"));
        console.log(`基礎ダメージは${baseDamage}です。`);
        const actualDamage = baseDamage - golem.defense < 0 ? 0 : baseDamage - golem.defense;
        console.log(`実際ダメージは${actualDamage}です。`);

        if (actualDamage <= 0) {
            playerHp = damagePlayer(golem.attack, playerHp);
            if (playerHp <= 0) {
                return {
                    message: "あなたはゴーレムに負けました！",
                    playerHp: null
                };
            }
            console.log(`あなたの残りHPは：${playerHp}`);
        } else {
            golem.hp = golem.hp - actualDamage < 0 ? 0 : golem.hp - actualDamage;
        }
    }

    console.log(`ゴーレムLv.${golemLevel + 1}を倒した！`);
    return {
        playerHp: playerHp,
        message: null
    };
}

function engageBattle(playerHp: number): ProgressResult {
    const golemLevel = getRandomNumber(10);
    const golem: Golem = {
        hp: golemLevel * 50 + 100,
        defense: golemLevel * 10 + 40,
        attack: golemLevel * 10 + 30
    };
    console.log(`ゴーレムLv.${golemLevel + 1}が現れた！`);
    return battleLoop(golemLevel, golem, playerHp);
}

export function gameLoop(playerHp: number): string {
    while (playerHp > 0) {
        console.log(`あなたのHP：${playerHp}`);

        switch (Number.parseInt(prompt("奥に進みますか？（１：奥に進む　０．帰る）＞") ?? "0")) {
            case 0:
                return "リレ〇ト！";
            default:
                {
                    const runBattle = () => {
                        const result = engageBattle(playerHp);
                        if (result.playerHp != null) {
                            playerHp = result.playerHp;
                        } else if (result.message != null) {
                            return result.message;
                        }
                        return null;
                    };

                    const result = runBattle();
                    if (result != null) {
                        console.log(result);
                        return "ゲームオーバー！";
                    }
                }
                break;
        }
    }

    return "ゲームオーバー！";
}